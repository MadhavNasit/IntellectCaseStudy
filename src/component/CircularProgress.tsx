import React, { FC, useEffect, useRef, useState } from 'react';
import { Animated, StyleSheet, View, ViewStyle } from 'react-native';
import Svg, { G, Linecap, Path } from 'react-native-svg';

export interface CircularProgressProps {
  /** Diameter of the circular progress (default: 140) */
  size?: number;
  /** Width of the progress stroke (default: 10) */
  strokeWidth?: number;
  /** Progress value from 0 to 100 (default: 50) */
  progress?: number;
  /** Color of the active progress stroke (default: '#0f1b27') */
  activeColor?: string;
  /** Color of the inactive background stroke (default: '#eae2df') */
  inActiveColor?: string;
  /** Animation duration in milliseconds (default: 500) */
  duration?: number;
  /** Arc sweep angle in degrees (default: 360 for full circle) */
  arcSweepAngle?: number;
  /** Rotation of the circular progress in degrees (default: 0) */
  rotation?: number;
  /** Stroke line cap style ('butt' | 'round' | 'square') (default: 'round') */
  strokeLinecap?: Linecap;
  /** Opacity of the active progress stroke (default: 1) */
  activeOpacity?: number;
  /** Opacity of the inactive background stroke (default: 1) */
  inActiveOpacity?: number;
  /** Style overrides for the container */
  containerStyle?: ViewStyle;
  /** Style overrides for the progress SVG */
  svgStyle?: ViewStyle;
  /** Style overrides for the inner content view */
  contentStyle?: ViewStyle;
  /** Children to be placed inside the progress circle */
  children?: React.ReactNode;
}

/**
 * CircularProgress - A customizable circular progress indicator component.
 *
 * This component renders a circular progress bar using SVG and Animated API.
 */
export const CircularProgress: FC<CircularProgressProps> = ({
  size = 140,
  strokeWidth = 10,
  progress = 50,
  activeColor = '#0f1b27',
  inActiveColor = '#eae2df',
  duration = 500,
  arcSweepAngle = 360,
  rotation = 0,
  strokeLinecap = 'round',
  activeOpacity = 1,
  inActiveOpacity = 1,
  containerStyle,
  svgStyle,
  contentStyle,
  children,
}) => {
  // Animated value to handle progress animation
  const progressAnim = useRef(new Animated.Value(progress)).current;
  const [animatedProgress, setAnimatedProgress] = useState(progress);

  useEffect(() => {
    // Animate progress changes
    Animated.timing(progressAnim, {
      toValue: progress,
      duration,
      useNativeDriver: false,
    }).start();

    // Listen to animation updates
    const listener = progressAnim.addListener(({ value }) => {
      setAnimatedProgress(value);
    });

    return () => {
      progressAnim.removeListener(listener);
    };
  }, [progress, duration, progressAnim]);

  // constants
  const padding = strokeWidth;
  const adjustedSize = size - padding * 2;
  const radius = adjustedSize / 2;
  const center = size / 2;

  /**
   * Creates an SVG arc path based on the given angle.
   * @param degree - The angle in degrees for the arc.
   * @returns {string} The SVG path data for the arc.
   */
  const createArcPath = (degree: number) => {
    // Manage full circle use case
    if (degree >= 360) {
      return `
        M ${center} ${center - radius}
        A ${radius} ${radius} 0 1 1 ${center} ${center + radius}
        A ${radius} ${radius} 0 1 1 ${center} ${center - radius}
      `;
    } else {
      const radians = (degree * Math.PI) / 180;
      const endX = center + radius * Math.cos(radians - Math.PI / 2);
      const endY = center + radius * Math.sin(radians - Math.PI / 2);
      const largeArcFlag = degree > 180 ? 1 : 0;
      return `
        M ${center} ${center - radius}
        A ${radius} ${radius} 0 ${largeArcFlag} 1 ${endX} ${endY}
      `;
    }
  };

  return (
    <View style={[{ width: size, height: size }, containerStyle]}>
      <Svg width={size} height={size} style={svgStyle}>
        <G rotation={rotation} origin={`${center}, ${center}`}>
          {/* Inactive background stroke */}
          <Path
            d={createArcPath(arcSweepAngle)}
            fill="none"
            stroke={inActiveColor}
            strokeWidth={strokeWidth}
            strokeLinecap={strokeLinecap}
            opacity={inActiveOpacity}
          />
          {/* Active progress stroke */}
          <Path
            d={createArcPath((arcSweepAngle * animatedProgress) / 100)}
            fill="none"
            stroke={activeColor}
            strokeWidth={strokeWidth}
            strokeLinecap={strokeLinecap}
            opacity={activeOpacity}
          />
        </G>
      </Svg>
      {/* Optional children inside the progress circle */}
      {!!children && (
        <View
          style={[
            styles.centerContent,
            {
              width: size * 0.7,
              height: size * 0.7,
              top: (size - size * 0.7) / 2,
              left: (size - size * 0.7) / 2,
            },
            contentStyle,
          ]}>
          {children}
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  centerContent: {
    position: 'absolute',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
