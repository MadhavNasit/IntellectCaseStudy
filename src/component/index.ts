export * from './Icon';
export * from './Screen';
export * from './Text';
export * from './Button';
export * from './Greeting';
export { default as ProgressBar } from './ProgressBar';
export * from './Card';
export * from './CircularProgress';
