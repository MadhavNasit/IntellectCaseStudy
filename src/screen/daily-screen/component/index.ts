export * from './SessionCard';
export { default as SessionList } from './SessionList';
