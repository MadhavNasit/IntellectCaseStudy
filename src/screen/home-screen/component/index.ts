export * from './CircularInsightItem';
export * from './PersonalInsightCard';
