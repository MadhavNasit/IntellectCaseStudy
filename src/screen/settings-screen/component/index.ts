export * from './LanguageSelector';
export * from './ThemeSelector';
