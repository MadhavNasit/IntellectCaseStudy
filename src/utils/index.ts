export * from './fontSize';
export * from './scale';
export * from './spacing';
