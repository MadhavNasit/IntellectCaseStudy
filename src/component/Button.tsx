import React, { useMemo } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
  TextStyle,
} from 'react-native';
import { useTheme } from '@react-navigation/native';

import { Text } from 'component';
import { Colors } from 'theme/theme.types';
import { fontSize, lineHeight, s, spacing, typography, vs } from 'utils';
import { TranslationKeys } from 'i18n';
import { TOptions } from 'i18next';

interface AppButtonProps extends TouchableOpacityProps {
  /** The text to be displayed on the button. */
  title?: string;
  /** Button text which is looked up via i18n. */
  titleTx?: TranslationKeys;
  /**
   * Optional options to pass to i18n. Useful for interpolation
   * as well as explicitly setting locale or translation fallbacks.
   */
  txOptions?: TOptions;
  /** The style preset to be applied ('primary', 'secondary', 'outline', 'link'). */
  preset?: 'primary' | 'secondary' | 'outline' | 'link';
  /** Custom styles to be applied to the button. */
  buttonStyle?: ViewStyle | ViewStyle[];
  /** Custom styles for the button's text. */
  buttonTextStyle?: TextStyle | TextStyle[];
  /** Optional component to be rendered on the left of the text. */
  leftComponent?: () => JSX.Element;
  /** Optional component to be rendered on the right of the text. */
  rightComponent?: () => JSX.Element;
  /** Boolean indicating whether the button is disabled or not. */
  disabled?: boolean;
}

/**
 * `Button` component that provides multiple button styles (primary, secondary, outline, link)
 * and the ability to customize its content with optional left and right components.
 */
export const Button: React.FC<AppButtonProps> = ({
  title,
  titleTx,
  txOptions,
  preset = 'primary',
  buttonStyle = {},
  buttonTextStyle = {},
  leftComponent,
  rightComponent,
  disabled = false,
  ...rest
}) => {
  const { colors } = useTheme();
  const styles = useMemo(() => makeStyles(colors), [colors]);

  const defaultStyle = useMemo(() => {
    let style: ViewStyle & { color?: string } = styles.button;

    // Determine default style based on the preset
    switch (preset) {
      case 'secondary':
        style = { ...style, ...styles.buttonSecondary };
        break;
      case 'outline':
        style = { ...style, ...styles.buttonOutline };
        break;
      case 'link':
        style = styles.link;
        break;
    }

    // Apply disabled styles if the button is disabled
    if (disabled) {
      style = { ...style, ...styles.buttonDisabled };
    }

    return style;
  }, [preset, disabled, styles]);

  return (
    <TouchableOpacity
      style={[defaultStyle, buttonStyle]}
      disabled={disabled}
      {...(preset === 'link' && {
        hitSlop: { top: 8, bottom: 8, left: 12, right: 12 },
      })}
      {...rest}>
      {leftComponent && leftComponent()}
      <Text
        text={title}
        tx={titleTx}
        txOptions={txOptions}
        numberOfLines={1}
        style={[styles.text, { color: defaultStyle.color }, buttonTextStyle]}>
        {title}
      </Text>
      {rightComponent && rightComponent()}
    </TouchableOpacity>
  );
};

/**
 * Creates a set of styles based on the current theme.
 *
 * @param colors - The theme colors to apply to the button styles.
 */
const makeStyles = (colors: Colors) =>
  StyleSheet.create({
    button: {
      flexDirection: 'row',
      backgroundColor: colors.primary,
      justifyContent: 'center',
      alignItems: 'center',
      color: colors.white,
      paddingVertical: vs(spacing.md),
      paddingHorizontal: s(spacing.lg),
      borderRadius: vs(30),
      columnGap: s(spacing.xs),
    } as ViewStyle,
    buttonSecondary: {
      backgroundColor: colors.secondary,
      color: colors.white,
    } as ViewStyle,
    buttonOutline: {
      backgroundColor: 'transparent',
      borderWidth: 2,
      borderColor: colors.primary,
      color: colors.text,
    } as ViewStyle,
    buttonDisabled: {
      opacity: 0.4,
    } as ViewStyle,
    link: {
      justifyContent: 'center',
      fontFamily: typography.regular,
      color: colors.text,
    } as ViewStyle,
    text: {
      fontFamily: typography.semiBold,
      fontSize: fontSize.h5,
      lineHeight: lineHeight[fontSize.h4],
    } as TextStyle,
  });
