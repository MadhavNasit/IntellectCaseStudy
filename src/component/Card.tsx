import React, { ComponentType, useMemo } from 'react';
import {
  StyleProp,
  StyleSheet,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';
import { useTheme } from '@react-navigation/native';
import { spacing, s, vs } from 'utils';
import { Colors } from 'theme/theme.types';

/**
 * Props for the Card component.
 */
export interface CardProps extends TouchableOpacityProps {
  /**
   * Children components to be rendered inside the card.
   */
  children?: React.ReactNode;
  /**
   * Style overrides for the card container.
   */
  style?: StyleProp<ViewStyle>;
}

/**
 * A reusable Card component that supports touch interactions.
 * It applies theme-based styles, including shadows and background color.
 *
 * - If `onPress` is provided, the card will be pressable.
 * - Accepts additional `TouchableOpacityProps` for customization.
 */
export const Card = ({
  children,
  style: styleOverride,
  ...rest
}: CardProps) => {
  // Access theme colors using useTheme hook
  const { colors } = useTheme();
  // Generate styles dynamically based on theme colors
  const styles = useMemo(() => makeStyles(colors), [colors]);

  // Determine if the card should be pressable
  const isPressable = !!rest.onPress;
  const Wrapper = (isPressable ? TouchableOpacity : View) as ComponentType<
    TouchableOpacityProps | ViewProps
  >;

  return (
    <Wrapper {...rest} style={[styles.card, styleOverride]}>
      {/* Render child components inside the card */}
      {children}
    </Wrapper>
  );
};

const makeStyles = (colors: Colors) =>
  StyleSheet.create({
    card: {
      borderWidth: 1,
      borderColor: colors.border,
      borderRadius: s(spacing.sm),
      paddingHorizontal: s(spacing.md),
      paddingVertical: vs(spacing.md),
      backgroundColor: colors.card,
    } as ViewStyle,
  });
