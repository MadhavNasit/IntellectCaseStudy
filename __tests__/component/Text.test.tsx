import React from 'react';
import { render } from '@testing-library/react-native';
import { Text } from 'component';
import { fontSize } from 'utils';
import {
  mockColors,
  mockTranslation,
  mockUseTheme,
  mockUseTranslation,
} from '../testUtils';

describe('Text Component', () => {
  beforeEach(() => {
    mockUseTranslation();
    mockUseTheme();
  });

  it('matches snapshot with text prop', () => {
    const { toJSON } = render(<Text text="Hello World" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with translated text', () => {
    const { toJSON } = render(<Text tx="common.appName" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with children text', () => {
    const { toJSON } = render(<Text>Some Child Text</Text>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with size prop', () => {
    const { toJSON } = render(<Text size="h1">Header Text</Text>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with custom styles', () => {
    const { toJSON } = render(
      <Text style={{ color: mockColors.text }} text="Styled Text" />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders text passed via the `text` prop', () => {
    const { getByText } = render(<Text text="Hello World" />);
    expect(getByText('Hello World')).toBeTruthy();
  });

  it('renders translated text when `tx` is provided', () => {
    const { getByText } = render(<Text tx="common.appName" />);
    expect(getByText(mockTranslation)).toBeTruthy();
  });

  it('renders text passed via children', () => {
    const { getByText } = render(<Text>Some Child Text</Text>);
    expect(getByText('Some Child Text')).toBeTruthy();
  });

  it('applies styles based on the `size` prop', () => {
    const { getByText } = render(<Text size="h1">Header Text</Text>);
    const textElement = getByText('Header Text');
    const style = textElement.props.style[1]; // h1 style from the makeStyles function
    expect(style.fontSize).toBe(fontSize.h1); // or other assertions based on your expected style
  });

  it('applies custom styles passed via the `style` prop', () => {
    const { getByText } = render(
      <Text style={{ color: mockColors.text }} text="Styled Text" />,
    );
    const textElement = getByText('Styled Text');
    const style = textElement.props.style[0];
    expect(style.color).toBe(mockColors.text);
  });
});
