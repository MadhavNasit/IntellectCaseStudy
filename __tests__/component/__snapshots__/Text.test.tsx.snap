// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component matches snapshot with children text 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
        "fontFamily": "Poppins-Regular",
      },
      {
        "fontSize": 14,
        "lineHeight": 20,
      },
      undefined,
    ]
  }
>
  Some Child Text
</Text>
`;

exports[`Text Component matches snapshot with custom styles 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
        "fontFamily": "Poppins-Regular",
      },
      {
        "fontSize": 14,
        "lineHeight": 20,
      },
      {
        "color": "#000000",
      },
    ]
  }
>
  Styled Text
</Text>
`;

exports[`Text Component matches snapshot with size prop 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
        "fontFamily": "Poppins-Regular",
      },
      {
        "fontFamily": "Poppins-SemiBold",
        "fontSize": 35,
        "lineHeight": 49,
      },
      undefined,
    ]
  }
>
  Header Text
</Text>
`;

exports[`Text Component matches snapshot with text prop 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
        "fontFamily": "Poppins-Regular",
      },
      {
        "fontSize": 14,
        "lineHeight": 20,
      },
      undefined,
    ]
  }
>
  Hello World
</Text>
`;

exports[`Text Component matches snapshot with translated text 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
        "fontFamily": "Poppins-Regular",
      },
      {
        "fontSize": 14,
        "lineHeight": 20,
      },
      undefined,
    ]
  }
>
  IntellectCaseStudy
</Text>
`;
